doggos/knowledge/rule.py

```diff
@@ -1,24 +1,105 @@
-from typing import Dict, Tuple, List
+from typing import Dict, NoReturn, List, NewType
 
 from doggos.fuzzy_sets.membership.membership_degree import MembershipDegree
-from doggos.knowledge.antecedent import Antecedent
+from doggos.knowledge import Antecedent, Clause
 from doggos.knowledge.consequents.consequent import Consequent
+from doggos.knowledge.consequents.takagi_sugeno_consequent import TakagiSugenoConsequent
+from doggos.knowledge.consequents.mamdani_consequent import MamdaniConsequent
 
 
 class Rule:
+    """
+    Class used to represent a fuzzy rule,
+    that is used in fuzzy logic systems to infer an output based on input variables.
+    It consists from an antecedent(premise) and a consequent.
+    https://en.wikipedia.org/wiki/Fuzzy_rule
+
+    Attributes
+    --------------------------------------------
+
+    __antecedent : Antecedent
+        Antecedent of the rule. Has getter and setter.
+
+    __consequent : Consequent
+        Consequent of the rule. Has getter and setter.
+
+    Methods
+    --------------------------------------------
+    output(self, features: Dict[str, float], measures: Dict[str, float] = None) -> OutputType:
+        Calculate firing value and output of rule, then return output of type that is depending on type of consequent.
+
+    Examples:
+    --------------------------------------------
+
+    """
+
     def __init__(self, antecedent: Antecedent, consequent: Consequent):
-        self.__firing_value: MembershipDegree = None
-        self.__rule_output: List[MembershipDegree] or float = None
-        self.__antecedent = antecedent
-        self.__consequent = consequent
+        """
+        Create a new rule with defined antecedent and consequent.
+
+        :param antecedent: antecedent of the rule
+        :param consequent: consequent of the rule
+        """
+        self.__antecedent = None
+        self.__consequent = None
+        if isinstance(antecedent, Antecedent):
+            self.__antecedent = antecedent
+        if isinstance(consequent, Consequent):
+            self.__consequent = consequent
```

Comment:
If antecedent or consequent are not instances of Antecedent or Consequent raise an exception.